src/mmc.c

```diff
@@ -217,36 +224,134 @@ int mmc_init(unsigned int id)
 	}
 
 	if (!retries) {
-		SERIAL_ERR(ERR_MMC_INIT);
+		SERIAL_ERR(ERR_MMC_SEND_OP_COND);
 		return -1;
 	}
 
+	bool is_sd = 1;
+
+	if (ret) {
+		if (ret == ERR_MMC_TIMEOUT) {
+			SERIAL_PUTS("trying mmc\n");
+
+			uint32_t ocr;
+			uint32_t arg = 0;
+			for (retries = 1000; retries; retries--)
+			{
+				ret = mmc_cmd(id, CMD_SEND_OP_COND, arg, 0x0, MSC_RESPONSE_R3, resp);
+
+				if (ret)
+					break;
+
+				ocr = (resp[0] >> 8) | (resp[1] << 8) | (resp[2] << 24);
+
+				if (ocr & MMC_OCR_BUSY)
+					break;
+
+				arg = ocr | MMC_OCR_HCS;
+
+				udelay(1000);
+			}
+
+			if (!retries || ret) {
+				int err;
+				if (!retries)
+					err = ERR_MMC_SEND_OP_COND_TOUT;
+				else
+					err = ERR_MMC_SEND_OP_COND_ERR;
+				SERIAL_ERR(err);
+				return -1;
+			}
+
+			SERIAL_PUTS_ARGH("OCR = ", ocr, "\n");
+
+			is_sd = false;
+		}
+		else
+		{
+			SERIAL_ERR(ERR_MMC_NTOUT);
+			return -1;
+		}
+	}
+
 	/* try to get card id */
-	mmc_cmd(id, CMD_ALL_SEND_CID, 0, 0x0, MSC_RESPONSE_R2, resp);
-	mmc_cmd(id, CMD_SEND_RCA, 0, 0x0, MSC_RESPONSE_R6, resp);
+	ret = mmc_cmd(id, CMD_ALL_SEND_CID, 0, 0x0, MSC_RESPONSE_R2, resp);
+
+	if (ret) {
+		SERIAL_ERR(ERR_MMC_CID);
+		return -1;
+	}
+
+	// set address
+	rca = 1;
 
-	rca = ((resp[2] & 0x00FF) << 24) | ((resp[1] & 0xFF00) << 8);
+	ret = mmc_cmd(id, CMD_SEND_RCA, rca << 16, 0x0, MSC_RESPONSE_R6, resp);
```

Comment:
You changed from calling CMD_SEND_RCA with arg == 0x0, to arg == 0x10000. Is that expected?

Reply:
yes it is, rca address is in high half of register

Reply:
Doesn't change the fact that in the SD case, the RCA value changed.

Reply:
yes it's expected, I set 0x01 addr in both cases
looks like 0x00 isnt valid in MMC case.

Reply:
OK, the SD specs does say that the argument is just "stuff bits": https://www.programmersought.com/article/19534548292/#6.cmd3